docs/developer/decision-records/2025-07-25-multiple-participant-identifiers/README.md

```diff
@@ -0,0 +1,114 @@
+to be considered for both the resolution of the own participant ID and the extraction of the counter-party ID.
+
+## Approach
+
+> Note: throughout this DR, we will use the term `protocol` as the key for all operations utilizing the
+> `DataspaceProfileRegistry` for backwards compatibility. This will be renamed to `profile` in the future.
+
+### Resolution of own participant ID
+
+Currently, the participant ID is determined through one setting and will be used all throughout the connector. We will
+add a participant ID to the `DataspaceProfileContext`, which allows assigning a specific identifier to a specific
+context. The `DataspaceProfileContextRegistry` will provide a new method for getting the participant ID for a given
+protocol:
+
+```java
+@NotNull
+String getParticipantId(String protocol);
+```
+
+The previously used participant ID will remain the default, i.e. it will be added as the participant ID to all default
+profiles registered at the `DataspaceProfileContextRegistry`. If no dedicated participant ID is defined for a given
+protocol, the method will return `null`:
+
+```java
+public String getParticipantId(String protocol) {
+    return profiles().stream()
+            .filter(it -> it.name().equals(protocol))
+            .map(DataspaceProfileContext::participantId)
+            .findAny()
+            .orElse(null);
+}
+```
+
+In all places where the participant ID is used (namely during creation of a catalog or a contract agreement), an
+instance of `DataspaceProfileContextRegistry` will be added, and the registry's `getParticipantId` method will be
+used where currently the participant ID is inserted.
+
+### Extraction of counter-party ID
+
+Right now, extraction of the counter-party ID from a `ClaimToken` is done in the `ParticipantAgentServiceImpl`, either
+through a `ParticipantAgentServiceExtension` or, as a fallback solution, by looking for a specific claim that can be
+defined through configuration.
+
+The extraction will be pulled up into the `ProtocolTokenValidatorImpl` to separate extraction of the ID from the
+extraction of any other participant attributes, which will still be handled by the `ParticipantAgentServiceExtensions`.
+The extracted ID will then be passed to the `ParticipantAgentService` as a parameter.
+
+As ID extraction should depend on the `DataspaceProfileContext`, a function for extracting the participant ID from a
+`ClaimToken` will be added to the `DataspaceProfileContext`. For this function, we add a new interface:
+
+```java
+@FunctionalInterface
+@ExtensionPoint
+public interface ParticipantIdExtractionFunction extends Function<ClaimToken, String> { }
+```
+
+Accordingly, we add another method to the `DataspaceProfileContextRegistry` for obtaining the ID extraction function
+for a given profile context:
+
+```java
+ParticipantIdExtractionFunction getIdExtractionFunction(String protocol);
+```
```

Comment:
This is for read, shouldn't also have a way to register `ParticipantIdExtractionFunction` for a specific profile?

Reply:
Andrea mentioned under his last comment that the default function should only be functional for default profiles, which I think makes sense. Therefore I changed the DR a bit to include the function into the `DataspaceProfileContext`, thus no new registration method would be needed. Wdyt?

Nevertheless, I could probably make it a bit more explicit that the function would be added to the `DataspaceProfileContext`, as this is easily missable right now I think 😄

Reply:
Make sense